packages/@glimmer/runtime/lib/vm/update.ts

```diff
@@ -196,189 +188,207 @@ export class TryOpcode extends BlockOpcode implements ExceptionHandler {
-  public map = new Map<unknown, BlockOpcode>();
-  public artifacts: IterationArtifacts;
   public tag: Tag;
+  public children!: ListItemOpcode[];
 
   private lastIterated: Revision = INITIAL;
   private _tag: UpdatableTag;
+  private opcodeMap = new Map<unknown, ListItemOpcode>();
+  private marker: SimpleComment | null = null;
+
+  protected readonly bounds!: LiveBlockList;
 
   constructor(
     state: ResumableVMState<InternalVM>,
     runtime: RuntimeContext,
-    bounds: LiveBlock,
-    children: LinkedList<UpdatingOpcode>,
-    artifacts: IterationArtifacts
+    bounds: LiveBlockList,
+    children: ListItemOpcode[],
+    private iterableRef: IterableReference
   ) {
     super(state, runtime, bounds, children);
-    this.artifacts = artifacts;
     let _tag = (this._tag = createUpdatableTag());
-    this.tag = combine([artifacts.tag, _tag]);
+    this.tag = combine([iterableRef.tag, _tag]);
+  }
+
+  initializeChild(opcode: ListItemOpcode) {
+    this.opcodeMap.set(opcode.key, opcode);
   }
 
   didInitializeChildren(listDidChange = true) {
-    this.lastIterated = valueForTag(this.artifacts.tag);
+    this.lastIterated = valueForTag(this.tag);
 
     if (listDidChange) {
-      updateTag(this._tag, combineSlice(this.children));
+      updateTag(this._tag, combineTagged(this.children));
     }
   }
 
   evaluate(vm: UpdatingVM) {
-    let { artifacts, lastIterated } = this;
+    let { iterableRef, lastIterated } = this;
 
-    if (!validateTag(artifacts.tag, lastIterated)) {
+    if (!validateTag(iterableRef.tag, lastIterated)) {
       let { bounds } = this;
       let { dom } = vm;
 
-      let marker = dom.createComment('');
+      let marker = (this.marker = dom.createComment(''));
       dom.insertAfter(
         bounds.parentElement(),
         marker,
         expect(bounds.lastNode(), "can't insert after an empty bounds")
       );
 
-      let target = new ListRevalidationDelegate(this, marker);
-      let synchronizer = new IteratorSynchronizer({ target, artifacts, env: vm.env });
-
-      synchronizer.sync();
+      this.sync();
 
       this.parentElement().removeChild(marker);
+      this.marker = null;
     }
 
     // Run now-updated updating opcodes
     super.evaluate(vm);
   }
 
-  vmForInsertion(nextSibling: Option<SimpleNode>): InternalVM<JitOrAotBlock> {
-    let { bounds, state, runtime } = this;
+  private sync() {
+    let { iterableRef, opcodeMap: itemMap, children } = this;
+
+    let item = iterableRef.next();
+    let currentOpcodeIndex = 0;
+
+    this.children = this.bounds.boundList = [];
+
+    while (item !== null) {
+      let opcode = children[currentOpcodeIndex];
+      let { key } = item;
+
+      if (opcode !== undefined && opcode === item.key) {
+        this.retainItem(opcode, item);
+        currentOpcodeIndex++;
+      } else if (itemMap.has(key)) {
+        let itemOpcode = itemMap.get(key)!;
+
+        if (itemOpcode.seen === true) {
+          this.moveItem(itemOpcode, item, opcode);
```

Comment:
will it support "cheap" 2 items swap? ref: https://github.com/glimmerjs/glimmer-vm/pull/945

Reply:
It should be exactly the same algorithm, just simplified. If it looks different, please let me know.

Reply:
@pzuraq looks like updated tests not counting `moving/retained` nodes count,  but it's quite expensive DOM operation, see https://github.com/glimmerjs/glimmer-vm/pull/1114/files#diff-6a84e01f254cabe1d50370a1616b9797L253
could we have minimal check for it?